### PR TITLE
refactor(experimental): add options package

### PR DESCRIPTION
**Pull request by lorisleiva**
This PR adds a new package that allows us to manage and encode/decode Rust-like `Option` types.

```ts
type Option<T> = Some<T> | None;
type Some<T> = { __option: "Some"; value: T };
type None = { __option: "None" };
```

This is particularly useful when generating clients from data structures defined in Rust. Whilst one could define an `Option<T>` type as `T | null` (currently serializable using the `getNullableCodec` method), this would loose information when using nested options. For instance, `Option<Option<T>>` would resolve to `T | T | null` which would resolve to `T | null` which is equivalent to `Option<T>`. In other words, `Some(None)` and `None` would not be distinguishable.

This new package provides everything we need when dealing with Rust-like `Option` types in JavaScript.

This is imported from the Umi framework as it is required when generating clients using Kinobi. See [Umi's Option documentation](https://github.com/metaplex-foundation/umi/blob/main/docs/helpers.md#options) for a more details description.

### CODE REVIEW

**Comment by lorisleiva** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1676** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1676" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1676?utm_source=stack-comment).

File: packages/codecs-data-structures/src/nullable.ts
Comment (steveluscher) [inline packages/codecs-data-structures/src/nullable.ts:1]: Were all these changes supposed to be in another PR?

File: packages/options/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+This package allows us to manage and serialize Rust-like Option types in JavaScript. It can be used standalone, but it is also exported as part of the Solana JavaScript SDK [`@solana/web3.js@experimental`](https://github.com/solana-labs/solana-web3.js/tree/master/packages/library).
+
+## Types
+
+TODO
+
+## Functions
+
+TODO
```
Comment (steveluscher) [inline packages/options/README.md:25]: I'm guilty of this too. We need to catch all the READMEs up, soon.

File: packages/options/src/option.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * @see {@link Option}
+ */
+export const some = <T>(value: T): Option<T> => ({ __option: 'Some', value });
+
+/**
+ * Creates a new {@link Option} of type `T` that has no value.
+ *
+ * @see {@link Option}
+ */
+export const none = <T>(): Option<T> => ({ __option: 'None' });
```
Comment (steveluscher) [inline packages/options/src/option.ts:42]: How would you feel about literally returning the same (frozen) object?
```suggestion
const NONE = Object.freeze({ __option: 'None' });
export const none = <T>(): Option<T> => NONE;
```

…then adding tests that `expect(none()).toBe(none())`?

File: packages/options/src/option.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * @see {@link Option}
+ */
+export type Some<T> = { __option: 'Some'; value: T };
+
+/**
+ * Represents an option of type `T` that has no value.
+ *
+ * @see {@link Option}
+ */
+export type None = { __option: 'None' };
```
Comment (steveluscher) [inline packages/options/src/option.ts:28]: These _really, really_ need to be `Readonly`.
```suggestion
/**
 * Represents an option of type `T` that has a value.
 *
 * @see {@link Option}
 */
export type Some<T> = Readonly<{ __option: 'Some'; value: T }>;

/**
 * Represents an option of type `T` that has no value.
 *
 * @see {@link Option}
 */
export type None = Readonly<{ __option: 'None' }>;
```

File: packages/options/src/option.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Whether the given {@link Option} is a {@link Some}.
+ */
+export const isSome = <T>(option: Option<T>): option is Some<T> => option.__option === 'Some';
+
+/**
+ * Whether the given {@link Option} is a {@link None}.
+ */
+export const isNone = <T>(option: Option<T>): option is None => option.__option === 'None';
```
Comment (steveluscher) [inline packages/options/src/option.ts:63]: Then…
```suggestion
export const isNone = <T>(option: Option<T>): option is None => option === NONE;
```

File: packages/options/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+    "$schema": "https://json.schemastore.org/tsconfig",
+    "display": "@solana/options",
+    "extends": "tsconfig/base.json",
+    "include": ["src"],
+    "compilerOptions": {
+        "lib": ["DOM", "ES2020", "ES2022.Error"]
```
Comment (steveluscher) [inline packages/options/tsconfig.json:7]: Copypasta? Eliminate items from this array until typechecking fails, and then put back only that which is needed to make it pass.

**Review by steveluscher — APPROVED** [review]
giphy-7

File: packages/options/src/option.ts (hunk 2 above)
Comment (lorisleiva) [reply packages/options/src/option.ts:42]: I'm happy to make that change but couldn't this create a dual-package hazard if an application ends up with multiple instances of this library (either because it's dependency chains ends up requiring both ESM and CJS or slightly different versions)? I've been trying to avoid adding top-level state for that purpose. Is this a valid concern here?

File: packages/options/README.md (hunk 1 above)
Comment (lorisleiva) [reply packages/options/README.md:25]: Yeah don't worry I'll make sure I tackle all these codec READMEs soon.

File: packages/options/src/option.ts (hunk 2 above)
Comment (lorisleiva) [reply packages/options/src/option.ts]: I've applied this suggestion but definitely think it's worth talking about it. 🙂
